src/OrleansRuntime/Catalog/Catalog.cs

```diff
@@ -567,130 +514,176 @@ private void SetupActivationInstance(ActivationData result, string grainType, st
+                }
+            }
+        }
 
-                    switch (initStage)
-                    {
-                        case 1: // failed to RegisterActivationInGrainDirectory
-                            
-                            ActivationAddress target = null;
-                            Exception dupExc;
-                            // Failure!! Could it be that this grain uses single activation placement, and there already was an activation?
-                            if (Utils.TryFindException(ex, typeof (DuplicateActivationException), out dupExc))
-                            {
-                                target = ((DuplicateActivationException) dupExc).ActivationToUse;
-                                CounterStatistic.FindOrCreate(StatisticNames.CATALOG_ACTIVATION_DUPLICATE_ACTIVATIONS)
-                                    .Increment();
-                            }
-                            activation.ForwardingAddress = target;
-                            if (target != null)
+        /// <summary>
+        /// Recover from a failed attempt to initialize a new activation.
+        /// </summary>
+        /// <param name="activation">The activation which failed to be initialized.</param>
+        /// <param name="initStage">The initialization stage at which initialization failed.</param>
+        /// <param name="registrationResult">The result of registering the activation with the grain directory.</param>
+        /// <param name="exception">The exception, if present, for logging purposes.</param>
+        private void RecoverFailedInitActivation(ActivationData activation, ActivationInitializationStage initStage,
+            ActivationRegistrationResult registrationResult, Exception exception = null)
+        {
+            ActivationAddress address = activation.Address;
+            lock (activation)
+            {
+                activation.SetState(ActivationState.Invalid);
+                try
+                {
+                    UnregisterMessageTarget(activation);
+                }
+                catch (Exception exc)
+                {
+                    logger.Warn(ErrorCode.Catalog_UnregisterMessageTarget4,
+                        string.Format("UnregisterMessageTarget failed on {0}.", activation), exc);
+                }
+
+                switch (initStage)
+                {
+                    case ActivationInitializationStage.Register: // failed to RegisterActivationInGrainDirectory
+
+                        // Failure!! Could it be that this grain uses single activation placement, and there already was an activation?
+                        activation.ForwardingAddress = registrationResult.ExistingActivationAddress;
+                        if (activation.ForwardingAddress != null)
+                        {
+                            CounterStatistic
+                                .FindOrCreate(StatisticNames.CATALOG_ACTIVATION_DUPLICATE_ACTIVATIONS)
```

Comment:
`public static readonly StatisticName CATALOG_ACTIVATION_DUPLICATE_ACTIVATIONS                               = new StatisticName("Catalog.Activation.DuplicateActivations");`

Potential names? `Catalog.Activation.RegistrationRaces`, `Catalog.Activation.ConcurrentRegistrationAttempts`? I prefer the second one, it's precise.